src/Operation/ReplaceOne.php

```diff
@@ -113,12 +113,12 @@ public function __construct(string $databaseName, string $collectionName, $filte
             }
 
             $replacement = $options['codec']->encodeIfSupported($replacement);
-        }
 
-        // Psalm's assert-if-true annotation does not work with unions, so
-        // assert the type manually instead of using is_document
-        // See https://github.com/vimeo/psalm/issues/6831
-        assert(is_array($replacement) || is_object($replacement));
+            // Psalm's assert-if-true annotation does not work with unions, so
+            // assert the type manually instead of using is_document
+            // See https://github.com/vimeo/psalm/issues/6831
+            assert(is_array($replacement) || is_object($replacement));
```

Comment:
Similar to my comments on InsertOne, I think we should apply the conditional encoding _before_ we validate `$replacement`. `encodeIfSupported()` is meant to accept anything, so I presume we can just pass whatever the user sends us.

Thinking about this more abstractly, this is somewhat related to my previous thread about middleware or decorating operations (https://github.com/mongodb/mongo-php-library/pull/1140#discussion_r1272520344), which lead to the creation of [PHPLIB-1210](https://jira.mongodb.org/browse/PHPLIB-1210). Applying a codec is something I envision as happening before the operation itself, although I understand we probably want to embed it in the operation class to avoid code duplication. And in some cases where we work with cursors (e.g. change streams), we have no choice but to handle it within.